sql/core/src/main/java/org/apache/spark/sql/sources/v2/PartitioningSupport.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2;
+
+/**
+ * A mix in interface for `DataSourceV2Reader` and `DataSourceV2Writer`. Users can implement this
+ * interface to specify the partition columns of the reader/writer, to improve performance.
```

Comment:
a data source may have metadata and keep the partitioning/bucketing information, but it may not have metadata and need to do expensive inference to retrieve partitioning/bucketing information. By mixing `BucketingSupport` and `PartitioningSupport`, a data source can get the partitioning/bucketing information directly, to avoid potential expensive inference.

Reply:
> Ideally we'd be able to put this a level lower

+1, I think it's doable to ask the parquet data source to talk with the metastore and get the partitioning/bucketing information.